### PR TITLE
Add moa and compound annotation

### PR DESCRIPTION
**Pull request by wxicu**
**PR Checklist**

-   [x] Referenced issue is linked #176 
-   [x] If you've fixed a bug or added code that should be tested, add tests!
-   [x] Documentation in `docs` is updated

**Description of changes**
Add compound and moa annotation `MoaMetaData` and `CompoundMetaData`
Improve `CelllineMetaData`: add new parameter `show` for each annotating function to print out only thr given number of unmatched identifiers.

**Technical details**
Compound annotation: fetch the compound annotation including the first synonym, cid and canonical smiles via pubchempy. query identifiers can be either CID or the name. 
moa annotation: fetch moa annotation sourced from clue.io. query id should be the name of the perturbagens. Annotated moa is also filtered on an optional parameter `target`.
Lookup objects for two new objects are also implemented.

**Additional context**

### CODE REVIEW

File: pertpy/tools/_metadata/_cell_line.py [hunk 1]
```diff
@@ -258,14 +260,23 @@ def annotate_cell_lines(
...
-                    f"[bold blue]There are {identifier_num_all} identifiers in `adata.obs`."
-                    f"However, {len(not_matched_identifiers)} identifiers can't be found in the cell line annotation,"
-                    "leading to the presence of NA values for their respective metadata.\n",
-                    "Please check again: ",
-                    *not_matched_identifiers,
-                    sep="\n- ",
-                )
+                if isinstance(show, str):
+                    if show != "all":
+                        raise ValueError("Only a non-positive value or all is accepted.")
```
Comment (Zethson) [inline pertpy/tools/_metadata/_cell_line.py:265]:
```suggestion
                        raise ValueError("Only a non-negative value or 'all' is accepted.")
```

non-negative, right?

File: pertpy/tools/_metadata/_cell_line.py [hunk 2]
```diff
@@ -258,14 +260,23 @@ def annotate_cell_lines(
...
+                    print(
+                        f"[bold blue]There are {identifier_num_all} identifiers in `adata.obs`."
+                        f"However, {len(not_matched_identifiers)} identifiers can't be found in the cell line annotation,"
+                        "leading to the presence of NA values for their respective metadata.\n",
+                        "Please check again: ",
+                        *not_matched_identifiers[:show],
+                        sep="\n- ",
+                    )
+                else:
+                    raise ValueError("Only 'all' or a non-positive value is accepted.")
```
Comment (Zethson) [inline pertpy/tools/_metadata/_cell_line.py:279]:
```suggestion
                    raise ValueError("Only 'all' or a non-negative value is accepted.")
```
?

File: pertpy/tools/_metadata/_cell_line.py [hunk 3]
```diff
@@ -394,15 +407,24 @@ def annotate_bulk_rna_expression(
...
-                f"[bold yellow]Following {len(not_matched_identifiers)} identifiers can't be found in bulk RNA expression data, "
-            )
-            print(
-                "leading to the presence of NA values for their respective metadata. Please check again: ",
-                *not_matched_identifiers,
-                sep="\n- ",
-            )
+            if isinstance(show, str):
+                if show != "all":
+                    raise ValueError("Only a non-positive value or all is accepted.")
```
Comment (Zethson) [inline pertpy/tools/_metadata/_cell_line.py:412]:
```suggestion
                    raise ValueError("Only a non-negative value or all is accepted.")
```

File: pertpy/tools/_metadata/_cell_line.py [hunk 4]
```diff
@@ -486,13 +510,24 @@ def annotate_protein_expression(
...
-            print(
-                f"[bold yellow]There are {identifier_num_all} identifiers in `adata.obs`. "
-                f"However {len(not_matched_identifiers)} identifiers can't be found in the protein expression data, "
-                "leading to the presence of NA values for their respective metadata. Please check again: ",
-                *not_matched_identifiers,
-                sep="\n- ",
-            )
+            if isinstance(show, str):
+                if show != "all":
+                    raise ValueError("Only a non-positive value or all is accepted.")
```
Comment (Zethson) [inline pertpy/tools/_metadata/_cell_line.py:515]:
```suggestion
                    raise ValueError("Only a non-negative value or all is accepted.")
```

File: pertpy/tools/_metadata/_cell_line.py [hunk 5]
```diff
@@ -548,14 +585,26 @@ def annotate_from_gdsc(
...
         if len(not_matched_identifiers) > 0:
-            print(
-                f"[bold yellow]Following {len(not_matched_identifiers)} identifiers can not be found in the drug response data for GDSC{gdsc_dataset},"
-                "leading to the presence of NA values for their respective metadata. Please check it again:",
-                *not_matched_identifiers,
-                sep="\n- ",
-            )
+            if isinstance(show, str):
+                if show != "all":
+                    raise ValueError("Only a non-positive value or all is accepted.")
```
Comment (Zethson) [inline pertpy/tools/_metadata/_cell_line.py:593]:
```suggestion
                    raise ValueError("Only a non-negative value or all is accepted.")
```

File: pertpy/tools/_metadata/_compound.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+from pertpy.data._dataloader import _download
+
+from ._look_up import LookUp
+
+if TYPE_CHECKING:
+    from anndata import AnnData
+
+
+class CompoundMetaData:
+    """Utilities to fetch metadata for compound."""
```
Comment (Zethson) [inline pertpy/tools/_metadata/_compound.py:21]:
```suggestion
    """Utilities to fetch metadata for compounds."""
```

File: pertpy/tools/_metadata/_compound.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        Args:
+            adata: The data object to annotate.
+            query_id: The column of `.obs` with compound identifiers. Defaults to "perturbation".
+            query_id_type: The type of compound identifiers, name or cid. Defaults to "name".
+            show: The number of unmatched identifiers to print, can be either non-negative values or "all". Defaults to 5.
+            copy: Determines whether a copy of the `adata` is returned. Defaults to False.
+
+        Returns:
+            Returns an AnnData object with compound annotation.
+
```
Comment (Zethson) [inline pertpy/tools/_metadata/_compound.py:47]:
```suggestion
```

File: pertpy/tools/_metadata/_compound.py [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+        """
+        if copy:
+            adata = adata.copy()
+
+        if query_id not in adata.obs.columns:
+            raise ValueError(f"The requested query_id {query_id} is not in `adata.obs`. \n" "Please check again. ")
+
+        query_dict = {}
+        not_matched_identifiers = []
+        for p in adata.obs[query_id].dropna().astype(str).unique():
```
Comment (Zethson) [inline pertpy/tools/_metadata/_compound.py:57]: What is `p`? Use a more descriptive name

File: pertpy/tools/_metadata/_compound.py [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+                "However, none of them could be found.\n"
+                "The annotation process has been halted.\n"
+                "To resolve this issue, please call the `CompoundMetaData.lookup()` function to create a LookUp object.\n"
+                "By using the `LookUp.compound()` method. "
+            )
+
+        if len(not_matched_identifiers) > 0:
+            if isinstance(show, str):
+                if show != "all":
+                    raise ValueError("Only a non-positive value or all is accepted.")
```
Comment (Zethson) [inline pertpy/tools/_metadata/_compound.py:87]:
```suggestion
                    raise ValueError("Only a non-negative value or 'all' is accepted.")
```

I'll stop pointing them out from now but you get the idea

File: pertpy/tools/_metadata/_look_up.py [hunk 10]
```diff
@@ -340,3 +376,63 @@ def available_genes_annotation(
...
+        if target_list is not None:
+            targets = self.moa_meta.target.astype(str).apply(lambda x: x.split("|"))
+            all_targets = [t for tl in targets for t in tl]
+            identifier_num_all = len(target_list)
+            not_matched_identifiers = list(set(target_list) - set(all_targets))
+            print(f"{len(not_matched_identifiers)} molecular targets are not found in the metadata.")
+            print(f"{identifier_num_all - len(not_matched_identifiers)} molecular targets are found! ")
+
+
+def available_compound(
```
Comment (Zethson) [inline pertpy/tools/_metadata/_look_up.py:409]:
```suggestion
def available_compounds(
```

Plural reads nicer I think?

Comment (Zethson) [reply pertpy/tools/_metadata/_look_up.py:409]: We should then use the plural across all functions where it reads better

File: tests/tools/_metadata/test_compound.py [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+        obs = obs.set_index(pd.Index([str(i) for i in range(NUM_GENES)]))
+        obs.index.rename("index", inplace=True)
+        obs["perturbation"] = (
+            ["AG-490"] * NUM_CELLS_PER_ID
+            + ["Iniparib"] * NUM_CELLS_PER_ID
+            + ["TAK-901"] * NUM_CELLS_PER_ID
+            + ["Quercetin"] * NUM_CELLS_PER_ID
+        )
+
+        var_data = {"gene_name": ["gene" + str(i) for i in range(1, NUM_GENES + 1)]}
```
Comment (Zethson) [inline tests/tools/_metadata/test_compound.py:40]:
```
var_data = {"gene_name": [f"gene{i}" for i in range(1, NUM_GENES + 1)]}
var = pd.DataFrame(var_data).set_index("gene_name", drop=False).rename_axis("index")
```

is a bit more concise

File: tests/tools/_metadata/test_compound.py [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+class TestMetaData:
+    pt_metadata = pt.tl.CompoundMetaData()
+
+    @pytest.fixture
+    def adata(self) -> AnnData:
+        rng = np.random.default_rng(1)
+        X = rng.standard_normal((NUM_CELLS, NUM_GENES))
+        X = np.where(X < 0, 0, X)
+
+        cell_line = {
```
Comment (Zethson) [inline tests/tools/_metadata/test_compound.py:23]:
```
cell_line = pd.DataFrame({
    "DepMap_ID": ["ACH-000016", "ACH-000049", "ACH-001208", "ACH-000956"] * NUM_CELLS_PER_ID,
    "perturbation": ["AG-490", "Iniparib", "TAK-901", "Quercetin"] * NUM_CELLS_PER_ID
}, index=[str(i) for i in range(NUM_GENES)])

obs = pd.concat([cell_line], axis=1)
obs.index.rename("index", inplace=True)
```

Maybe something like this?

File: tests/tools/_metadata/test_moa.py [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+class TestMetaData:
+    pt_metadata = pt.tl.MoaMetaData()
+
+    @pytest.fixture
+    def adata(self) -> AnnData:
+        rng = np.random.default_rng(1)
+        X = rng.standard_normal((NUM_CELLS, NUM_GENES))
+        X = np.where(X < 0, 0, X)
+
+        cell_line = {
```
Comment (Zethson) [inline tests/tools/_metadata/test_moa.py:23]: Maybe also simplify this - see above

**Review by Zethson — COMMENTED** [review]
Thank youuu!

1. Could you please always make the "available database" names links? Also to any of the older cell line metadata databases, please
2. All of these `show` checks seem pretty repetitive. Is there a way to extract this as a helper function? You just need to also parameterize the type of data (protein, [...]), but that's easy.
3. I think that some of the code can be reduced and simplified. It's a bit verbose.

File: pertpy/tools/_metadata/_look_up.py (hunk 10 above)
Comment (wxicu) [reply pertpy/tools/_metadata/_look_up.py]: currently only two are in plural (compound + cell line). what about bulk RNA expression, protein expression and moa? i keep it as singular first

File: pertpy/tools/_metadata/_cell_line.py [hunk 14]
```diff
@@ -196,6 +196,7 @@ def annotate_cell_lines(
         reference_id: str = "DepMap_ID",
         cell_line_information: list[str] | None = None,
         cell_line_source: Literal["DepMap", "Cancerrxgene"] = "DepMap",
+        show: int | str = 5,
```
[the remark this answers is not in the document]
Comment (wxicu) [reply pertpy/tools/_metadata/_cell_line.py]: verbose? show_warning?

File: pertpy/tools/_metadata/_look_up.py [hunk 15]
```diff
@@ -340,3 +376,63 @@ def available_genes_annotation(
...
+        self,
+        query_id_list: list[str] | None = None,
+        target_list: list[str] | None = None,
+    ) -> None:
+        """A brief summary of MoA annotation.
+
+        Args:
+            query_id_list: A list of unique perturbagens to test the number of matched ones present in the metadata. Defaults to None.
+            target_list: A list of unique molecular targets to test the number of matched ones present in the metadata. Defaults to None.
+
```
Comment (Zethson) [inline pertpy/tools/_metadata/_look_up.py:390]:
```suggestion
```

File: pertpy/tools/_metadata/_look_up.py [hunk 16]
```diff
@@ -340,3 +376,63 @@ def available_genes_annotation(
...
+        query_id_list: A list of unique compounds to test the number of matched ones present in the metadata. Defaults to None.
+        query_id_type: The type of compound identifiers, name or cid. Defaults to "name".
+    """
+    if self.type != "compound":
+        raise ValueError("This is not a LookUp object specific for CompoundData!")
+    if query_id_list is not None:
+        identifier_num_all = len(query_id_list)
+        not_matched_identifiers = []
+
+        for p in query_id_list:
```
Comment (Zethson) [inline pertpy/tools/_metadata/_look_up.py:426]: What is 'p`?

**Review by Zethson — COMMENTED** [review]
Could you please check the types of many of your parameters again? I think they could be a bit more general. Maybe `Iterable` instead of `List` for some etc.
Also make the behavior clear if `None` is an option. Sometimes you're not describing that.

File: pertpy/metadata/_cell_line.py [hunk 17]
```diff
@@ -294,6 +297,9 @@ def annotate_cell_lines(
                 # We will subset the original metadata dataframe correspondingly and add them to the AnnData object.
                 # Again, redundant information will be removed.
                 if reference_id not in cell_line_information:
+                    # Sequence doesnt have append function, turn into list
```
Comment (Zethson) [inline pertpy/metadata/_cell_line.py:300]: Well, in that case, we could actually require it to be `list`. I didn't see this ufff. You can change it now, but you don't have to

File: pertpy/metadata/_look_up.py [hunk 18]
```diff
@@ -340,3 +383,62 @@ def available_genes_annotation(
...
+            raise ValueError("This is not a LookUp object specific for MoaMetaData!")
+        if query_id_list is not None:
+            identifier_num_all = len(query_id_list)
+            not_matched_identifiers = list(set(query_id_list) - set(self.moa_meta.pert_iname))
+            print(f"{len(not_matched_identifiers)} perturbagens are not found in the metadata.")
+            print(f"{identifier_num_all - len(not_matched_identifiers)} perturbagens are found! ")
+
+        if target_list is not None:
+            targets = self.moa_meta.target.astype(str).apply(lambda x: x.split("|"))
+            all_targets = [t for tl in targets for t in tl]
```
Comment (Zethson) [inline pertpy/metadata/_look_up.py:408]: Not the prettiest or clearest naming. What is `tl`?

**Review by Zethson — COMMENTED** [review]
What are you deleting the whole cell line test and where did your new tests go?

We can merge this then!

**Pull request merged**